scheduler/scheduler.go

```diff
@@ -526,29 +629,54 @@ func (driver *MesosSchedulerDriver) Start() (mesos.Status, error) {
 	driver.setStatus(mesos.Status_DRIVER_RUNNING)
 	driver.setStopped(false)
 
-	log.Infoln("Mesos scheduler driver started with PID=", driver.self.String())
+	log.Infof("Mesos scheduler driver started with PID=%v", driver.self)
 
+	// start the detector AFTER we have a self pid from the messenger, otherwise things get ugly.
+	// route detector callbacks over the messenger bus, maintaining serial callback execution.
+	listener := detector.AsMasterChanged(func(m *mesos.MasterInfo) {
+		driver.messenger.Route(context.TODO(), driver.self, &mesos.InternalMasterChangeDetected{
+			Master: m,
+		})
+	})
+	log.V(1).Infof("starting master detector %T: %+v", driver.masterDetector, driver.masterDetector)
+	driver.masterDetector.Detect(listener)
+
+	log.V(2).Info("master detector started")
 	return driver.Status(), nil
 }
 
-// authenticate?
-//TODO(jdef) perhaps at some point in the future this will get pushed down into
-//the messenger layer (e.g. to use HTTP-based authentication). We'd probably still
-//specify the callback.Handler here, along with the user-selected authentication
-//provider. Perhaps in the form of some messenger.AuthenticationConfig.
-
-// authenticate using the installed authentication plugin.
-func (driver *MesosSchedulerDriver) authenticate() error {
-	log.Infoln("Authenticating with master.")
+// authenticate against the spec'd master pid using the configured authenticationProvider.
+// the authentication process is canceled upon either closing of the authenticating chan, or
+// else because it timed out (authTimeout).
+//
+// TODO(jdef) perhaps at some point in the future this will get pushed down into
+// the messenger layer (e.g. to use HTTP-based authentication). We'd probably still
+// specify the callback.Handler here, along with the user-selected authentication
+// provider. Perhaps in the form of some messenger.AuthenticationConfig.
+//
+func (driver *MesosSchedulerDriver) authenticate(pid *upid.UPID, authenticating <-chan struct{}) error {
+	log.Infof("authenticating with master %v", pid)
 	ctx, cancel := context.WithTimeout(context.Background(), authTimeout)
-	defer cancel()
 	handler := &CredentialHandler{
-		pid:        driver.MasterPid,
-		client:     driver.messenger.UPID(),
+		pid:        pid,
+		client:     driver.self,
 		credential: driver.credential,
 	}
 	ctx = auth.WithLoginProvider(ctx, *authProvider)
-	return auth.Login(ctx, handler)
+	ch := make(chan error, 1)
+	go func() { ch <- auth.Login(ctx, handler) }()
+	select {
+	case <-ctx.Done():
+		<-ch
+		return ctx.Err()
+	case <-authenticating:
+		cancel()
```

Comment:
I have been seeing this cancel() method in several go files now, where is that defined ?